### ISSUE TITLE
Visiblity after an illegal move

### ISSUE THREAD

**Issue by Jester565** [Jester565/SpyNinjaGame#21]
Try this...
1. Look up
2. Try to move out of bounds

...and now you can't see?
The visibililty was reset too early.  This was a major problem that I had to fix in this commit of DevAhead 9dc2bccbb8420d166e521e7a7d8d034b5964e056 after learning that moving after an illegal move would be allowed.
Look at GameEngine.playerMove() and notice that I actually call Grid.setToInvisible() inside of the method.  You can fix it a different way but just remember that the visibility of the board must be reset before moving anything.

**Comment by Jester565** [comment]
And Francisco, the reason checkMove() exists is because of this issue.  In the commit where I fix this, I wrote: 
"Grid.move(DIRECTION, int, int)'s behavior was split into two methods.
The first part is Grid.checkMoveStatus(DIRECTION direction, int x, int
y), which will no longer move the GameObject but will return the
moveStatus.  Grid.move(DIRECTION moveDirection, int x, int y) now
handles actually moving the GameObject (be careful no checks are
performed by this method)."

Basically what I'm trying to say is that move(DIRECTION, int, int) should have no need to perform checks because it should already be checked by Grid.checkMoveStatus.

**Comment by jtara1** [comment]
I'll work on having the visibility not reset when an `ILLEGAL` move is done.

**Comment by jtara1** [comment]
@Jester565 

~~is this what you wanted?~~ 743db8dc3cd4f3fe36461890e77af8e902c727c2

edit: what did I do, now the Spy is not being drawn on the grid? I need to go through this again ...

edit2: reverted that commit because that was a really bad commit

**Comment by jtara1** [comment]
Really tired right now, I'll look at this in the morning.

**Comment by Jester565** [comment]
The problem: There's an array in Grid that is supposed to keep track of what elements in the Grid are visible called [visibleObjects](https://github.com/Jester565/SpyNinjaGame/blob/master/SpyGame/src/edu/cpp/cs/cs141/final_proj/Grid.java#L61).  The array is added to in the function [setAsVisible()](https://github.com/Jester565/SpyNinjaGame/blob/master/SpyGame/src/edu/cpp/cs/cs141/final_proj/Grid.java#L77-L100).  When the element it is setting as visible is null, it will put a GameObject called [VisibleMark](https://github.com/Jester565/SpyNinjaGame/blob/master/SpyGame/src/edu/cpp/cs/cs141/final_proj/VisibleMark.java) in that spot.  If the element it is setting is not null, it will call [GameObject.setVisiblity(true)](https://github.com/Jester565/SpyNinjaGame/blob/master/SpyGame/src/edu/cpp/cs/cs141/final_proj/GameObject.java#L93-L96).  So now we have all of our visibleObjects and the Grid will draw properly with visibility and everything.  The issue is when we try to remove our visibleObjects, after all we don't want to leave VisibleMarks and objects setToVisible behind us.  The removing is done by the method [Grid.setToInvisible()](https://github.com/Jester565/SpyNinjaGame/blob/master/SpyGame/src/edu/cpp/cs/cs141/final_proj/Grid.java#L105-L119).  This method is the destroyer of worlds!  It is very dangerous.  It will just iterate over visibleObjects and will remove the objects with its same x and y values as VisibleMarks.  REMEMBER its removing based on x and y values!  That is why Spy was being removed from the Grid!  Let me clarify.

We add a VisibleMark to visibleObjects.  We then move the Spy so he is standing on top of a VisibleMark.  We then call setToInvisible() which looks at the VisibleMark's x and y value and removes whatever GameObject is at that position (in this case it was the Spy standing on top of the VisibleMark).

Solutions:
Method A: You need to resetVisibility() before you move anything!  Anything at all!  Here is a link to a commit I did that solved that very problem 9dc2bccbb8420d166e521e7a7d8d034b5964e056.  It basically just makes sure that a move is not illegal before resetting visibility in the [playerTurn()](https://github.com/Jester565/SpyNinjaGame/blob/9dc2bccbb8420d166e521e7a7d8d034b5964e056/SpyGame/src/edu/cpp/cs/cs141/final_proj/GameEngine.java#L246-L262) method.  The only down side is that you have to split the moveCheck from the move, but we have to do that anyways... so yeah.

Method B: Allow move to handle removing visibility.  Here is a link to a commit where I removed that functionality, but you can see all the code deleted 0e5a5cc9f47762c0f3f215b85d517a2afd2d8ce1

**Comment by jtara1** [comment]
**TIL**: type "#21" in commit message creates link to this issue (#21), and typing "close #21" automatically closes it.

So I didn't mean to close this issue, but it should solve the problem.
